### DISCUSSION TITLE
无缝交互

### DISCUSSION THREAD

**Discussion by halogp** [kbengine/kbengine#1545] [Q&A]
这个引擎是支持无缝地图的，那在逻辑代码里面，一个玩家可以直接修改另一个玩家的数据吗，比如给周围的一个人加buff。
因为另一个人可能是一个ghost，那数据的修改会不会被real同步过来的数据给覆盖掉。

**Comment by huhubis** [comment]
对于ghost那是可能覆盖的

**Reply by halogp** [reply]
像这种一个物件修改另一个物件的属性或者方法，需要什么规则来保证开发流程正确吗，因为另一个物件很有可能是ghost。

**Reply by kibozh** [reply]
ghost的属性是只读的，所以你改不了，你只能通过他的real去改，也就不会有你说的被real覆盖掉的问题了

**Reply by ve-i-uj** [reply]
Ghost properties can be changed by calling a remote method through another component (`CellViaBase`). For example, there is an entity `entity1` in Python code on Cellapp and `entity1` is a ghost. You can change a property of a real entity by calling a remote method on that entity's ghost. The method will be called on a real entity, in which you can already change the property value.

```python
entity1.base.cell.change_my_property('new_value')
```

`base.cell` is `CellViaBase`, a call to a remote cell method via the base component. Baseapp knows where the real entity is and calls the method on the real entity. Through a method you can change a property of a real entity.

In `entity_defs` you need to add a `change_my_property` method to the `entity1` entity class. This method should not be `<Exposed/>`.

Here is an example based on the demo, where I call the cell method of the Avatar entity through the base component. I added a `change_my_property` method to the avatar in the `Avatar.def` file and implemented the method in `cell/Avatar.py`

```console
leto@leto-Latitude-5591:~/2PeopleCompany/REPOS/shedu$ telnet localhost 50001
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
password:pwd123456

[cellapp@python ~] >>> 
[cellapp@python ~] >>> a1, *_ = [e for e in KBEngine.entities.values() if e.__class__.__name__ == 'Avatar']
[cellapp@python ~] >>> a1
Avatar object at 0x7fcacbb60b18.

[cellapp@python ~] >>> a1.base.cell
CellViaBase id:2144, utype:2, component=baseapp[8001], addr: 172.20.0.9:39083.

[cellapp@python ~] >>> a1.base.cell.change_my_property
EntityRemoteMethod object at 0x7fcacbb6f978.

[cellapp@python ~] >>> a1.base.cell.change_my_property('123')
[cellapp@python ~] >>> 
```
